Improve Governance module functionality (#460)
LiskHQ/lips: proposals/lip-0074.md

|   **Name**      |  **Type**   |   **Modifiable**  |**Initial Mainchain Value**|**Description**         |
|-----------------|-------------|-------------------|----------------------------------|------------------------|
|`tokenIdTreasury`|   `bytes`   |     No            | `TOKEN_ID_LSK`                   |Token ID of the tokens stored in the treasury and handled by funding proposals.|
|`voteDuration`   | `uint32`    |       No          |   `241,920 = 28 * 24 * 3600 // BLOCK_TIME`                         |Length of the vote period in blocks. As clarified in [Rationale](#voting-on-a-proposal), `voteDuration <= LOCKING_PERIOD_STAKING`, where `LOCKING_PERIOD_STAKING` is defined in [LIP 0057][posModule].|

> |`quorumDuration` | `uint32`    |       Yes          |   `120,960 = 14 * 24 * 3600 // BLOCK_TIME`                         |Length of the quorum period in blocks. After this period the quorum is checked.|

Reviewer: If the goal is to configure these durations based on time, not on number of block passed, it would be better to simply define them as `28 * 24 * 3600 // BLOCK_TIME` and `14 * 24 * 3600 // BLOCK_TIME` and remove the calculated values `241,920` and `120,960`.

The current specification is misleading IMO, because it assumes that the `BLOCK_TIME` is `10`, while in fact it is a configurable value.

Author: I think the current version is fine. The LIP suggests initial values for the Mainchain, so it is fair to assume that `BLOCK_TIME = 10`. Moreover I mention this parameter explicitly, so if in the future we decide to change `BLOCK_TIME` value, a simple search will also indicate that the Governance module parameters should be updated accordingly.

    elif type == PROPOSAL_TYPE_CONFIG_UPDATE:
        proposalDataObj = decode(configUpdateProposalDataSchema, data)
        validateObjectSchema(configUpdateProposalDataSchema, proposalDataObj)

>         proposalDataObj.moduleName.validateConfigUpdate(proposalDataObj.paramName, proposalDataObj.value)

Author: Changed as suggested.

    if type == PROPOSAL_TYPE_FUNDING:
        # assume that the data is valid
        dataObj = decode(fundingProposalDataSchema, data)
        # lock donated treasury tokens together with minted treasury tokens
        treasuryDonations = Token.getAvailableBalance(ADDRESS_TREASURY, configStore.tokenIdTreasury)
        if (treasuryDonations != 0):

>             Token.lock(ADDRESS_TREASURY, MODULE_NAME_GOVERNANCE, configStore.tokenIdTreasury, treasuryDonations)

Reviewer 2: I see you decided to lock the donated tokens, so they become undistinguishable from the minted tokens. I can't think of any reason why this shouldn't be so, and it seems like a fine solution to me. 
I suppose adding a comment to the Treasury part of the Rationale stating that users can freely donate to treasury is unnecessary, right?

Author: Have added a sentence to the rationale.

        call updateConfig(dataObj.paramName, dataObj.value) method of module dataObj.moduleName
```

#### validateConfigUpdate

The function raises an exception if the proposed config update is invalid.

```python
def validateConfigUpdate(name: string, value: bytes):

>     type = type of value with name in configSchema

Reviewer 2: To me this looks more readable:
```suggestion
    type = string indicating the type of name property in configSchema
```

Or maybe this?
```suggestion
    type = configSchema["properties"][name]["type"]
```

But probably it doesn't need to be changed, because then it would need to be changed in `updateConfig`, too... just a thinking-aloud-kind-of-comment, feel free to resolve.

Author: Agree, the current version is not very readable. I have changed to your first suggestion.

    valueDecoded = validateAndDecode(value, type)
    if name == "quorumPercentage":
        if valueDecoded > 1000000:
            raise Exception("Quorum percentage must be in parts-per-million")
    if name == "quorumDuration":
        if valueDecoded > configStore.voteDuration:

>             raise Exception("Quorum duration can not be greater than vote duration")

Reviewer 2: Should here be some `else` indicating what happens if `name` is none of the ones stated above?

Author: I have added a check that name is a valid config name, so now the function should be fine.
